File: 20210128-java-lift-kata/src/test/java/codingdojo/LiftSystemTest.java
package codingdojo;

import org.junit.jupiter.api.Test;

import static java.util.List.of;
import static org.approvaltests.Approvals.verify;

public class LiftSystemTest {

    @Test
    public void a_lift_should_not_move_when_there_is_no_request_for_it() {
        Lift liftA = new Lift("A", 0);
        LiftSystem lifts = new LiftSystem(of(0, 1), of(liftA), of());
        lifts.tick();
        verify(new LiftSystemPrinter().print(lifts));
    }

    @Test
    public void a_lift_should_move_to_requested_floor_and_open_doors_in_order_to_fulfill_request() {
        Lift liftA = new Lift("A", 0);
        LiftSystem lifts = new LiftSystem(of(0, 1), of(liftA), of(new Call(1, Direction.DOWN)));
        lifts.tick();
        verify(new LiftSystemPrinter().print(lifts));
    }
}

File: 202102225-kotlin-lift-kata/src/test/kotlin/codingdojo/LiftSystemTest.kt
package codingdojo

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class LiftSystemTest {

    @Test
    fun `a lift should move to requested floor`() {
        val liftA = Lift(id = "A", floor = 0, requests = listOf(1))
        val liftSystem = LiftSystem(floors = listOf(0, 1), lifts = listOf(liftA), calls = emptyList())

        liftSystem.tick()

        assertThat(liftA.getFloor()).isEqualTo(1)
    }

    @Test
    fun `a lift should not move when already in requested floor`() {
        val liftA = Lift(id = "A", floor = 0, requests = listOf(0))
        val liftSystem = LiftSystem(floors = listOf(0, 1), lifts = listOf(liftA), calls = emptyList())

        liftSystem.tick()

        assertThat(liftA.getFloor()).isEqualTo(0)
    }

    @Test
    fun `a lift should not move when doors are open`() {
        val liftA = Lift(id = "A", floor = 0, doorsOpen = true, requests = listOf(1))
        val liftSystem = LiftSystem(floors = listOf(0, 1), lifts = listOf(liftA), calls = emptyList())

        liftSystem.tick()

        assertThat(liftA.getFloor()).isEqualTo(0)
    }

    @Test
    fun `a lift should close the doors before moving`() {
        val liftA = Lift(id = "A", floor = 0, doorsOpen = true, requests = listOf(1))
        val liftSystem = LiftSystem(floors = listOf(0, 1), lifts = listOf(liftA), calls = emptyList())

        liftSystem.tick()

        assertThat(liftA.areDoorsOpen()).isFalse()
    }

    @Test
    fun `a lift should open doors after arriving to the requested floor`() {
        val liftA = Lift(id = "A", floor = 0, doorsOpen = false, requests = listOf(1))
        val liftSystem = LiftSystem(floors = listOf(0, 1), lifts = listOf(liftA), calls = emptyList())

        liftSystem.tick()
        assertThat(liftA.areDoorsOpen()).isFalse()
        assertThat(liftA.getFloor()).isEqualTo(1)

        liftSystem.tick()
        assertThat(liftA.areDoorsOpen()).isTrue()
    }
}
